File: code/backend/python/app/adapters/rest/match/turn_cycle_controller.py
"""Step 24 — FastAPI controller for the single-player turn cycle endpoints.

Endpoints (player, Bearer access token):
  POST /api/matches/{uuid_match}/start          -> 200 | 401 | 404 | 409
  POST /api/gameplay/{uuid_match}/action/pass   -> 200 | 401 | 404 | 409
  GET  /api/match/{uuid_match}/turn-sequence     -> 200 | 401 | 404
"""
import time

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

from app.core.models.match.turn_models import (
    PassResult,
    TurnCycleError,
    TurnSequenceResult,
)
from app.core.ports.match.turn_ports import TurnCyclePort

# HTTP status per error code (mirrors the Java controller).
_STATUS_BY_CODE = {
    TurnCycleError.MATCH_NOT_FOUND: 404,
    TurnCycleError.MATCH_NOT_STARTABLE: 409,
    TurnCycleError.NO_CHARACTERS_JOINED: 409,
    TurnCycleError.MATCH_NOT_RUNNING: 409,
    TurnCycleError.NOT_YOUR_TURN: 409,
}


def _error(code: str, message: str, http_status: int) -> JSONResponse:
    return JSONResponse(
        status_code=http_status,
        content={"error": code, "message": message, "timestamp": int(time.time() * 1000)},
    )


def _entry_to_camel(e) -> dict:
    return {
        "characterUuid": e.character_uuid,
        "idCharacter": e.id_character,
        "name": e.name,
        "priority": e.priority,
        "clock": e.clock,
        "status": e.status,
        "passCounter": e.pass_counter,
        "timestampStart": e.timestamp_start,
        "timestampEnd": e.timestamp_end,
    }


def _sequence_to_camel(r: TurnSequenceResult) -> dict:
    return {
        "matchUuid": r.match_uuid,
        "currentClock": r.current_clock,
        "status": r.status,
        "activeCharacterUuid": r.active_character_uuid,
        "queue": [_entry_to_camel(e) for e in r.queue],
    }


def _pass_to_camel(r: PassResult) -> dict:
    return {
        "matchUuid": r.match_uuid,
        "passedCharacterUuid": r.passed_character_uuid,
        "nextActiveCharacterUuid": r.next_active_character_uuid,
        "status": r.status,
    }


class TurnCycleController:
    def __init__(self, turn_cycle_port: TurnCyclePort):
        self.turn_cycle_port = turn_cycle_port
        self.router = APIRouter()
        self.router.add_api_route(
            "/api/matches/{uuid_match}/start", self.start, methods=["POST"]
        )
        self.router.add_api_route(
            "/api/gameplay/{uuid_match}/action/pass", self.pass_turn, methods=["POST"]
        )
        self.router.add_api_route(
            "/api/match/{uuid_match}/turn-sequence", self.turn_sequence, methods=["GET"]
        )

    def start(self, uuid_match: str, request: Request):
        user_uuid = getattr(request.state, "user_uuid", None)
        if not user_uuid:
            return _error("UNAUTHENTICATED", "User identity is missing", 401)
        try:
            result = self.turn_cycle_port.start_match(uuid_match, user_uuid)
        except TurnCycleError as exc:
            return _error(exc.code, exc.message, _STATUS_BY_CODE.get(exc.code, 400))
        return JSONResponse(status_code=200, content=_sequence_to_camel(result))

    def pass_turn(self, uuid_match: str, request: Request):
        user_uuid = getattr(request.state, "user_uuid", None)
        if not user_uuid:
            return _error("UNAUTHENTICATED", "User identity is missing", 401)
        try:
            result = self.turn_cycle_port.pass_turn(uuid_match, user_uuid)
        except TurnCycleError as exc:
            return _error(exc.code, exc.message, _STATUS_BY_CODE.get(exc.code, 400))
        return JSONResponse(status_code=200, content=_pass_to_camel(result))

    def turn_sequence(self, uuid_match: str, request: Request):
        user_uuid = getattr(request.state, "user_uuid", None)
        if not user_uuid:
            return _error("UNAUTHENTICATED", "User identity is missing", 401)
        try:
            result = self.turn_cycle_port.get_turn_sequence(uuid_match, user_uuid)
        except TurnCycleError as exc:
            return _error(exc.code, exc.message, _STATUS_BY_CODE.get(exc.code, 400))
        return JSONResponse(status_code=200, content=_sequence_to_camel(result))

File: code/backend/java/adapter-rest/src/main/java/games/paths/adapters/rest/controller/match/TurnCycleController.java
package games.paths.adapters.rest.controller.match;

import games.paths.adapters.rest.dto.PassTurnResponse;
import games.paths.adapters.rest.dto.TurnSequenceResponse;
import games.paths.core.port.match.TurnCyclePort;

import jakarta.servlet.http.HttpServletRequest;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * TurnCycleController - REST adapter for the single-player turn cycle (Step 24).
 *
 * <ul>
 *   <li>POST /api/matches/&#123;uuidMatch&#125;/start          — CREATED → RUNNING + build queue</li>
 *   <li>POST /api/gameplay/&#123;uuidMatch&#125;/action/pass   — voluntary turn pass</li>
 *   <li>GET  /api/match/&#123;uuidMatch&#125;/turn-sequence    — read the turn queue</li>
 * </ul>
 *
 * <p>See {@code documentation_v0/Step24_TurnCycleEngine.md}.</p>
 */
@RestController
public class TurnCycleController {

    private final TurnCyclePort turnCyclePort;

    public TurnCycleController(TurnCyclePort turnCyclePort) {
        this.turnCyclePort = turnCyclePort;
    }

    @PostMapping("/api/matches/{uuidMatch}/start")
    public ResponseEntity<Object> start(@PathVariable String uuidMatch, HttpServletRequest request) {
        String userUuid = userUuid(request);
        if (userUuid == null) {
            return error(HttpStatus.UNAUTHORIZED, "UNAUTHENTICATED", "User identity is missing");
        }
        try {
            return ResponseEntity.ok(TurnSequenceResponse.fromModel(
                    turnCyclePort.startMatch(uuidMatch, userUuid)));
        } catch (TurnCyclePort.TurnCycleException ex) {
            return error(mapStatus(ex.getCode()), ex.getCode().name(), ex.getMessage());
        }
    }

    @PostMapping("/api/gameplay/{uuidMatch}/action/pass")
    public ResponseEntity<Object> pass(@PathVariable String uuidMatch, HttpServletRequest request) {
        String userUuid = userUuid(request);
        if (userUuid == null) {
            return error(HttpStatus.UNAUTHORIZED, "UNAUTHENTICATED", "User identity is missing");
        }
        try {
            return ResponseEntity.ok(PassTurnResponse.fromModel(
                    turnCyclePort.passTurn(uuidMatch, userUuid)));
        } catch (TurnCyclePort.TurnCycleException ex) {
            return error(mapStatus(ex.getCode()), ex.getCode().name(), ex.getMessage());
        }
    }

    @GetMapping("/api/match/{uuidMatch}/turn-sequence")
    public ResponseEntity<Object> sequence(@PathVariable String uuidMatch, HttpServletRequest request) {
        String userUuid = userUuid(request);
        if (userUuid == null) {
            return error(HttpStatus.UNAUTHORIZED, "UNAUTHENTICATED", "User identity is missing");
        }
        try {
            return ResponseEntity.ok(TurnSequenceResponse.fromModel(
                    turnCyclePort.getTurnSequence(uuidMatch, userUuid)));
        } catch (TurnCyclePort.TurnCycleException ex) {
            return error(mapStatus(ex.getCode()), ex.getCode().name(), ex.getMessage());
        }
    }

    private static String userUuid(HttpServletRequest request) {
        String u = (String) request.getAttribute("userUuid");
        return (u == null || u.isBlank()) ? null : u;
    }

    private static ResponseEntity<Object> error(HttpStatus status, String code, String message) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("error", code);
        body.put("message", message);
        body.put("timestamp", System.currentTimeMillis());
        return ResponseEntity.status(status).body(body);
    }

    private static HttpStatus mapStatus(TurnCyclePort.TurnCycleException.Code code) {
        return switch (code) {
            case MATCH_NOT_FOUND -> HttpStatus.NOT_FOUND;
            case MATCH_NOT_STARTABLE, NO_CHARACTERS_JOINED, MATCH_NOT_RUNNING, NOT_YOUR_TURN ->
                    HttpStatus.CONFLICT;
        };
    }
}
